File: frontend/lib/model/judgement/userankpairjudgement/model/UseRankPairJudgement.ts
import UseRankPairInstance from "../../../instance/userankpairinstance/model/UseRankPairInstance";
import { IJudgement, IJudgementConstructor } from "../../model/IJudgement";
import UseRankPairJudgementDto from "../dto/UseRankPairJudgementDto";
import UseRankPairJudgementId from "./UseRankPairJudgementId";

export default class UseRankPairJudgement implements IJudgement {

    readonly id: UseRankPairJudgementId;

    readonly instance: UseRankPairInstance;
    
    readonly label: string;
    readonly comment: string;

    constructor(id: UseRankPairJudgementId, instance: UseRankPairInstance, label: string, comment: string) {
        this.id = id;
        this.instance = instance;
        this.label = label;
        this.comment = comment;
    }

    public getId(): UseRankPairJudgementId {
        return this.id;
    }

    public getInstance(): UseRankPairInstance {
        return this.instance;
    }

    public getLabel(): string {
        return this.label;
    }

    public getComment(): string {
        return this.comment;
    }

    public static fromDto(dto: UseRankPairJudgementDto): UseRankPairJudgement {
        return new UseRankPairJudgement(
            UseRankPairJudgementId.fromDto(dto.id),
            UseRankPairInstance.fromDto(dto.instance),
            dto.label,
            dto.comment
        );
    }

}

export class UseRankPairJudgementConstructor implements IJudgementConstructor {
    fromDto(dto: UseRankPairJudgementDto): UseRankPairJudgement {
        return UseRankPairJudgement.fromDto(dto);
    }
}

File: backend/src/main/java/de/garrafao/phitag/domain/judgement/userankpairjudgement/UseRankPairJudgement.java
package de.garrafao.phitag.domain.judgement.userankpairjudgement;

import de.garrafao.phitag.domain.annotator.Annotator;
import de.garrafao.phitag.domain.instance.userankpairinstances.UseRankPairInstance;
import de.garrafao.phitag.domain.judgement.IJudgement;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.Validate;

import javax.persistence.*;

@Entity
@Table(name = "phitaguserankpairjudgement")
@Getter
public class UseRankPairJudgement implements IJudgement {

    @Setter
    @EmbeddedId
    private UseRankPairJudgementId id;

    @MapsId("annotatorid")
    @ManyToOne
    private Annotator annotator;

    @MapsId("instanceid")
    @ManyToOne
    private UseRankPairInstance instance;

    @Setter
    @Column(name = "rank", nullable = false)
    private String  label;

    @Setter
    @Column(name = "comment", nullable = false)
    private String comment;

    UseRankPairJudgement() {
    }

    public UseRankPairJudgement(UseRankPairInstance instance, Annotator annotator, String label, String comment) {
        Validate.notNull(instance, "instanceData must not be null");
        Validate.notNull(annotator, "annotator must not be null");
        Validate.notNull(label, "label must not be null");
        Validate.notNull(comment, "comment must not be null");

        this.id = new UseRankPairJudgementId(instance.getId(), annotator.getId());

        this.instance = instance;
        this.annotator = annotator;

        this.label = label;
        this.comment = comment;
    }



    public UseRankPairInstance getUseRankPairInstance() {
        return instance;
    }

    @Override
    public String toString() {
        return "UseRankPairJudgement [id=" + id + ", label=" + label + ", comment=" + comment + "]";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        UseRankPairJudgement other = (UseRankPairJudgement) obj;
        return this.id.equals(other.getId());
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

}
